rust/arrow/src/array/equal.rs

```diff
@@ -418,6 +420,171 @@ fn value_offset_equal<T: Array + ListArrayOps>(this: &T, other: &T) -> bool {
     true
 }
 
+/// Trait for comparing arrow array with json array
+pub trait JsonEqual {
+    fn equals_json(&self, json: &[&Value]) -> bool;
+
+    fn equals_json_values(&self, json: &[Value]) -> bool {
+        let refs = json.iter().map(|v| v).collect::<Vec<&Value>>();
+
+        self.equals_json(&refs)
+    }
+}
+
+/// Implement array equals for numeric type
+impl<T: ArrowPrimitiveType> JsonEqual for PrimitiveArray<T> {
+    fn equals_json(&self, json: &[&Value]) -> bool {
+        if self.len() != json.len() {
+            return false;
+        }
+
+        let result = (0..self.len()).all(|i| match json[i] {
+            Value::Null => self.is_null(i),
+            v => self.is_valid(i) && Some(v) == self.value(i).into_json_value().as_ref(),
+        });
+
+        result
+    }
+}
+
+impl<T: ArrowPrimitiveType> PartialEq<Value> for PrimitiveArray<T> {
+    fn eq(&self, json: &Value) -> bool {
+        match json {
+            Value::Array(array) => self.equals_json_values(&array),
+            _ => false,
+        }
+    }
+}
+
+impl<T: ArrowPrimitiveType> PartialEq<PrimitiveArray<T>> for Value {
+    fn eq(&self, arrow: &PrimitiveArray<T>) -> bool {
+        match self {
+            Value::Array(array) => arrow.equals_json_values(&array),
+            _ => false,
+        }
+    }
+}
+
+impl JsonEqual for ListArray {
+    fn equals_json(&self, json: &[&Value]) -> bool {
+        if self.len() != json.len() {
+            return false;
+        }
+
+        let result = (0..self.len()).all(|i| match json[i] {
+            Value::Array(v) => self.is_valid(i) && self.value(i).equals_json_values(v),
+            Value::Null => self.is_null(i) || self.value_length(i) == 0,
+            _ => false,
+        });
+
+        result
+    }
+}
+
+impl PartialEq<Value> for ListArray {
+    fn eq(&self, json: &Value) -> bool {
+        match json {
+            Value::Array(json_array) => self.equals_json_values(json_array),
+            _ => false,
+        }
+    }
+}
+
+impl PartialEq<ListArray> for Value {
+    fn eq(&self, arrow: &ListArray) -> bool {
+        match self {
+            Value::Array(json_array) => arrow.equals_json_values(json_array),
+            _ => false,
+        }
+    }
+}
+
+impl JsonEqual for StructArray {
+    fn equals_json(&self, json: &[&Value]) -> bool {
+        if self.len() != json.len() {
+            return false;
+        }
+
+        let all_object = json.iter().all(|v| match v {
+            Object(_) | JNull => true,
+            _ => false,
+        });
+
+        if !all_object {
+            println!("Not all objects");
```

Comment:
nit: should we keep these `println!()` here?

Reply:
Used for debugging, removed.